File: legacy/src/game/pinball-knight/maze/modifiers.ts
/**
 * FLOOR MODIFIERS — the roguelike "this floor is different" layer.
 *
 * Archetypes change a floor's shape and themes change its furniture, but both
 * are on a fixed cycle: once you have seen 20 floors you have seen every
 * pairing. A modifier is rolled from the floor's own seed, so two runs that
 * reach depth 7 don't get the same depth 7. It is announced on the descent card
 * — an unannounced modifier reads as the game being buggy, not as a twist.
 *
 * Every field is a MULTIPLIER on a budget core.ts already computes, so a
 * modifier can never invent a mechanic the floor builder doesn't have; the
 * worst a bad roll can do is a floor that is too dark or too crowded. Nothing
 * here can affect connectivity.
 *
 * DOM- and three-free: tested in modifiers.test.ts.
 */
import { MAT_ICE, MAT_MUD, MAT_RUBBER, MAT_BRASS } from "../engine/surfaces";

export type ModifierId = "none" | "flooded" | "blackout" | "overcharged" | "gilded" | "collapsing" | "frozen" | "silted";

export interface FloorModifier {
  id: ModifierId;
  /** Shown on the descent card. Empty for "none" so the card stays clean. */
  label: string;
  flavour: string;
  /** Multipliers on the level budgets (1 = leave alone). */
  torchMult: number;
  partMult: number;
  hordeMult: number;
  hazardMult: number;
  trapdoorMult: number;
  /** Extra loot rolls beyond the level's normal armoury. */
  bonusItems: number;
  /**
   * Part kinds pushed to the FRONT of the theme's deal, biasing which furniture
   * the corridor pass reaches for first.
   */
  dealBias: string[];
  /**
   * What the floor is MADE OF (engine/surfaces.ts): relative weights over
   * MaterialId, stamped as patches by maze/surface-paint.ts.
   *
   * This is the first modifier field that changes PHYSICS rather than a budget.
   * It is still safe against the module's standing guarantee that "nothing here
   * can affect connectivity", because the painter only ever rewrites what a
   * tile is made of — it never moves one. An empty mix paints nothing.
   */
  surfaceMix: Record<number, number>;
  /** Roughly what fraction of the floor the mix covers, 0..1. */
  surfaceCoverage: number;
}

const BASE: Omit<FloorModifier, "id" | "label" | "flavour"> = {
  torchMult: 1,
  partMult: 1,
  hordeMult: 1,
  hazardMult: 1,
  trapdoorMult: 1,
  bonusItems: 0,
  dealBias: [],
  surfaceMix: {},
  surfaceCoverage: 0,
};

export const MODIFIERS: FloorModifier[] = [
  {
    ...BASE,
    id: "none",
    label: "",
    flavour: "",
  },
  {
    // Nothing brakes. The floor is a slip-and-slide and bumpers become the only
    // way to shed speed — which makes them read as safety rather than hazard.
    ...BASE,
    id: "flooded",
    label: "Flooded",
    flavour: "ankle-deep and slick · nothing here stops",
    dealBias: ["oil", "oil", "bumper"],
    partMult: 1.15,
    hazardMult: 1.2,
    // The flavour text has promised "nothing here stops" since this modifier
    // shipped; until surfaces existed the only thing backing it up was an oil
    // deal bias. Ice patches make the claim literally true underfoot.
    surfaceMix: { [MAT_ICE]: 1 },
    surfaceCoverage: 0.3,
  },
  {
    // Half-lit, double-paid. The classic risk/reward dark floor.
    ...BASE,
    id: "blackout",
    label: "Blackout",
    flavour: "the torches are out · something wanted them out",
    torchMult: 0.45,
    bonusItems: 2,
    hordeMult: 0.85,
    // The only modifier where the player cannot SEE the floor, so it is the
    // one where a surface has to be legible by ear and by feel instead. Brass
    // is exactly that: ordinary restitution, so it never surprises you, but
    // double combo ticks and a bright spark on contact — in the dark the walls
    // you bounce off become the map. A small mud fraction keeps it a gamble
    // rather than a gift, since a chain broken by a wall you never saw is the
    // risk half of a risk/reward floor.
    //
    // Explicitly NOT ice or heavy rubber: a surface that steers or throws you
    // is a surface you have to see coming, and stacking one on an unlit floor
    // is not difficulty, it is a coin flip.
    surfaceMix: { [MAT_BRASS]: 4, [MAT_MUD]: 1 },
    surfaceCoverage: 0.24,
  },
  {
    // A denser table AND a bigger horde — the floor that plays fastest.
    ...BASE,
    id: "overcharged",
    label: "Overcharged",
    flavour: "the machinery is running hot · everything hits harder",
    partMult: 1.4,
    hordeMult: 1.25,
    dealBias: ["bumper", "slingshot", "glove"],
    // Hot machinery = live rubber. Small patches, because rubber walls GAIN and
    // a floor that is mostly rubber is a floor with no brakes anywhere.
    surfaceMix: { [MAT_RUBBER]: 1 },
    surfaceCoverage: 0.16,
  },
  {
    // Treasure floor: the armoury is fat, and so is what's guarding it.
    ...BASE,
    id: "gilded",
    label: "Gilded",
    flavour: "gold in the cracks · and a queue for it",
    bonusItems: 3,
    hordeMult: 1.6,
    hazardMult: 0.8,
    // Brass bell-plate: the treasure floor is also the SCORING floor, since
    // every strike on it is worth double on the chain.
    surfaceMix: { [MAT_BRASS]: 1 },
    surfaceCoverage: 0.22,
  },
  {
    // The floor is coming apart: hatches and pits everywhere, few torches.
    // Falling costs you position, not your run, so this stays a pace modifier.
    ...BASE,
    id: "collapsing",
    label: "Collapsing",
    flavour: "the floor is giving way · keep moving",
    trapdoorMult: 2.2,
    hazardMult: 1.5,
    torchMult: 0.75,
    dealBias: ["pit", "trapdoor"],
    // "Keep moving" is a rule the floor should enforce, not just announce.
    // Rubble underfoot (sand drags, mud walls break the chain) is what makes
    // standing still expensive, and the scattered rubber slabs are the sprung
    // masonry of a floor coming apart — the one thing that will still throw
    // you clear of a hatch. Wide coverage because a collapsing floor that is
    // mostly intact is just a floor with some holes in it.
    surfaceMix: { [MAT_MUD]: 3, [MAT_RUBBER]: 1 },
    surfaceCoverage: 0.4,
  },
  {
    // SURFACE-LED. The whole floor is a skating rink: you keep every unit of
    // speed you build and almost none of your steering, so the route has to be
    // chosen before you commit to it rather than corrected mid-ride. Fewer
    // parts on purpose — the ICE is the obstacle, and stacking a dense table on
    // top of it makes an unreadable floor rather than a hard one.
    ...BASE,
    id: "frozen",
    label: "Frozen",
    flavour: "black ice wall to wall · pick your line early",
    partMult: 0.85,
    hazardMult: 0.9,
    surfaceMix: { [MAT_ICE]: 5, [MAT_MUD]: 1 },
    surfaceCoverage: 0.62,
  },
  {
    // SURFACE-LED, and the mirror of Frozen: mud kills chains, so a silted
    // floor is where combo scoring goes to die and raw clearing pays instead.
    // The bonus items are the compensation for a floor that refuses to be
    // played stylishly.
    ...BASE,
    id: "silted",
    label: "Silted",
    flavour: "silt over everything · the chain dies where it touches",
    bonusItems: 2,
    hordeMult: 0.9,
    surfaceMix: { [MAT_MUD]: 4, [MAT_BRASS]: 1 },
    surfaceCoverage: 0.45,
  },
];

const byId = new Map(MODIFIERS.map((m) => [m.id, m]));

export function modifierById(id: ModifierId): FloorModifier {
  return byId.get(id) ?? MODIFIERS[0];
}

/** The "no modifier" record — exported so callers can default without a lookup. */
export const NO_MODIFIER = MODIFIERS[0];

/**
 * Levels 1-2 never roll one: the opening floors are where a player is still
 * learning the base rules, and a twist there reads as the rules being unclear.
 */
export const MODIFIER_FROM_LEVEL = 3;
/** Chance a qualifying floor draws a modifier at all. */
export const MODIFIER_CHANCE = 0.45;

/**
 * Roll this floor's modifier. Draws ZERO rng values below MODIFIER_FROM_LEVEL
 * (early return) and one or two otherwise (one when the chance gate fails), so
 * the stream stays predictable for a given (run, level).
 */
export function rollModifier(level: number, rng: () => number): FloorModifier {
  if (level < MODIFIER_FROM_LEVEL) return NO_MODIFIER;
  if (rng() >= MODIFIER_CHANCE) return NO_MODIFIER;
  const pool = MODIFIERS.filter((m) => m.id !== "none");
  return pool[Math.floor(rng() * pool.length)];
}

File: Rust/crates/pk-core/src/maze/modifiers.rs
//! FLOOR MODIFIERS — the once-in-a-while twist, rolled from the floor's seed.
//!
//! Port of `legacy/src/game/pinball-knight/maze/modifiers.ts`, **roll only**.
//!
//! ## Why the roll ships before the effects
//!
//! A modifier scales content budgets — horde size, torches, part count, the
//! order `decorate` reaches for furniture — and none of that exists in Rust
//! yet. What DOES exist downstream of this function is the whole floor: it is
//! the first thing `authorFloor` draws, so every later draw on the floor sits
//! behind it. Get the roll wrong by one value and `growTrack` seeds a different
//! network, on a floor that is otherwise correctly ported.
//!
//! So the id and the DRAW COUNT are ported here and the multipliers land with
//! `decorate`. [`ModifierId`] carries the full table in order because the
//! second draw indexes it — a shortened pool picks a different modifier and,
//! worse, is invisible until content exists to be scaled.
//!
//! ⚠️ Verified against the fixture's `drawsBeforeTrack`, which is 0, 1, 2 or 3
//! across the corpus precisely because of the two early returns below.
//!
//! PORTS: `maze/modifiers.ts`

use crate::maze::CountingRng;

/// The table, in order. Index 0 is the "no modifier" record and is NOT in the
/// pool the roll indexes — see [`roll_modifier`].
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ModifierId {
    None,
    Flooded,
    Blackout,
    Overcharged,
    Gilded,
    Collapsing,
    Frozen,
    Silted,
}

/// Everything except `None`, in table order — what the second draw indexes.
pub const MODIFIER_POOL: [ModifierId; 7] = [
    ModifierId::Flooded,
    ModifierId::Blackout,
    ModifierId::Overcharged,
    ModifierId::Gilded,
    ModifierId::Collapsing,
    ModifierId::Frozen,
    ModifierId::Silted,
];

/// Levels 1-2 never roll one: the opening floors are where a player is still
/// learning the base rules, and a twist there reads as the rules being unclear.
pub const MODIFIER_FROM_LEVEL: i32 = 3;

/// Chance a qualifying floor draws a modifier at all.
pub const MODIFIER_CHANCE: f64 = 0.45;

/// Roll this floor's modifier.
///
/// Draws ZERO values below [`MODIFIER_FROM_LEVEL`] and one or two otherwise
/// (one when the chance gate fails), so the stream stays predictable for a
/// given (run, level) — and so `drawsBeforeTrack` varies across the corpus in a
/// way the fixture can pin.
pub fn roll_modifier(level: i32, rng: &mut CountingRng) -> ModifierId {
    if level < MODIFIER_FROM_LEVEL {
        return ModifierId::None;
    }
    if rng.next_f64() >= MODIFIER_CHANCE {
        return ModifierId::None;
    }
    // `Math.floor(rng() * n)` — the draw is in [0,1) so the index is in range,
    // but the floor is written out rather than cast blindly: `as usize` on a
    // negative would wrap, and this is the kind of line that gets copied.
    let k = (rng.next_f64() * MODIFIER_POOL.len() as f64).floor() as usize;
    MODIFIER_POOL[k.min(MODIFIER_POOL.len() - 1)]
}

#[derive(Clone, Debug, PartialEq)]
pub struct FloorModifier {
    pub id: ModifierId,
    pub label: &'static str,
    pub flavour: &'static str,
    pub torch_mult: f64,
    pub part_mult: f64,
    pub horde_mult: f64,
    pub hazard_mult: f64,
    pub trapdoor_mult: f64,
    pub bonus_items: i32,
    pub deal_bias: &'static [&'static str],
    pub surface_coverage: f64,
}

impl ModifierId {
    pub fn data(&self) -> FloorModifier {
        match self {
            ModifierId::None => FloorModifier {
                id: ModifierId::None,
                label: "",
                flavour: "",
                torch_mult: 1.0,
                part_mult: 1.0,
                horde_mult: 1.0,
                hazard_mult: 1.0,
                trapdoor_mult: 1.0,
                bonus_items: 0,
                deal_bias: &[],
                surface_coverage: 0.0,
            },
            ModifierId::Flooded => FloorModifier {
                id: ModifierId::Flooded,
                label: "Flooded",
                flavour: "ankle-deep and slick · nothing here stops",
                torch_mult: 1.0,
                part_mult: 1.15,
                horde_mult: 1.0,
                hazard_mult: 1.2,
                trapdoor_mult: 1.0,
                bonus_items: 0,
                deal_bias: &["oil", "oil", "bumper"],
                surface_coverage: 0.3,
            },
            ModifierId::Blackout => FloorModifier {
                id: ModifierId::Blackout,
                label: "Blackout",
                flavour: "the torches are out · something wanted them out",
                torch_mult: 0.45,
                part_mult: 1.0,
                horde_mult: 0.85,
                hazard_mult: 1.0,
                trapdoor_mult: 1.0,
                bonus_items: 2,
                deal_bias: &[],
                surface_coverage: 0.24,
            },
            ModifierId::Overcharged => FloorModifier {
                id: ModifierId::Overcharged,
                label: "Overcharged",
                flavour: "the machinery is running hot · everything hits harder",
                torch_mult: 1.0,
                part_mult: 1.4,
                horde_mult: 1.25,
                hazard_mult: 1.0,
                trapdoor_mult: 1.0,
                bonus_items: 0,
                deal_bias: &["bumper", "slingshot", "glove"],
                surface_coverage: 0.16,
            },
            ModifierId::Gilded => FloorModifier {
                id: ModifierId::Gilded,
                label: "Gilded",
                flavour: "gold in the cracks · and a queue for it",
                torch_mult: 1.0,
                part_mult: 1.0,
                horde_mult: 1.6,
                hazard_mult: 0.8,
                trapdoor_mult: 1.0,
                bonus_items: 3,
                deal_bias: &[],
                surface_coverage: 0.22,
            },
            ModifierId::Collapsing => FloorModifier {
                id: ModifierId::Collapsing,
                label: "Collapsing",
                flavour: "the floor is giving way · keep moving",
                torch_mult: 0.75,
                part_mult: 1.0,
                horde_mult: 1.0,
                hazard_mult: 1.5,
                trapdoor_mult: 2.2,
                bonus_items: 0,
                deal_bias: &["pit", "trapdoor"],
                surface_coverage: 0.4,
            },
            ModifierId::Frozen => FloorModifier {
                id: ModifierId::Frozen,
                label: "Frozen",
                flavour: "black ice wall to wall · pick your line early",
                torch_mult: 1.0,
                part_mult: 0.85,
                horde_mult: 1.0,
                hazard_mult: 0.9,
                trapdoor_mult: 1.0,
                bonus_items: 0,
                deal_bias: &[],
                surface_coverage: 0.62,
            },
            ModifierId::Silted => FloorModifier {
                id: ModifierId::Silted,
                label: "Silted",
                flavour: "silt over everything · the chain dies where it touches",
                torch_mult: 1.0,
                part_mult: 1.0,
                horde_mult: 0.9,
                hazard_mult: 1.0,
                trapdoor_mult: 1.0,
                bonus_items: 2,
                deal_bias: &[],
                surface_coverage: 0.45,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shallow_floors_draw_nothing() {
        for level in [1, 2] {
            let mut rng = CountingRng::new(1);
            assert_eq!(roll_modifier(level, &mut rng), ModifierId::None);
            assert_eq!(rng.draws(), 0, "level {level} must not touch the stream");
        }
    }

    #[test]
    fn a_qualifying_floor_draws_one_or_two() {
        // Seeds chosen only to exercise both branches; the VALUES are pinned by
        // the fixture's drawsBeforeTrack, not here.
        let mut both = (false, false);
        for seed in 0..64u32 {
            let mut rng = CountingRng::new(seed);
            roll_modifier(3, &mut rng);
            match rng.draws() {
                1 => both.0 = true,
                2 => both.1 = true,
                n => panic!("seed {seed}: {n} draws, wanted 1 or 2"),
            }
        }
        assert!(both.0 && both.1, "one seed range hit only one branch");
    }
}
